CL2-6617 Avoid inconsistent data check failures related to IdeasPhase validity

jinjagit opened the pull request:
## Checklist

- [ ] Added entry to changelog
<details>
<summary>More info</summary>
<br>
No changelog entry included, since this change aims to reduce our reporting of inconsistent data, and is thus of little direct benefit to users and difficult to explain in language that is understandable outside of our dev team. 
</details>

- [x] Tests
<details>
<summary>More info</summary>

### Acceptance test
Added test that asserts a related `IdeasPhase` remains valid after updating an ideation phase with ideas to a poll phase
(see changes in `back/spec/acceptance/phases_spec.rb`)

</details>

- [x] Prepared branch for code review
<details>
<summary>More info</summary>
</details>

## Links

- [Jira ticket](https://citizenlab.atlassian.net/browse/CL2-6617)

## What changes are in this PR?

We decided to remove the the `phase_is_ideation` validation from the `IdeasPhase` model.

This should avoid the related inconsistent data check failures seen in CI, as in these examples:
```
"issues": {
    "IdeasPhase_phase_id_phase_not_ideation": {
      "count": 45,
      "hosts": {
        "rijswijk-buiten.citizenlab.co": [
          "xxxxxxxxx-cbf0-47bb-a416-b7d82e876859",
          "xxxxxxxxx-553d-44b5-81d3-793e80707b1e",
          "xxxxxxxxx-474f-443e-b499-72c53f4cf648",
          "xxxxxxxxx-799f-44f8-af3b-270771cba317",
          "xxxxxxxxx-ad38-4026-ac7c-7e0ebec46b15"
        ],
        "denkmee.katwijk.nl": [
          "xxxxxxxxx-7274-45f9-b59d-698be01254c0"
        ], ...
```
Excerpt from [this CI run](https://app.circleci.com/pipelines/github/CitizenLabDotCo/citizenlab-ee/4767/workflows/1047a799-8912-49bc-81ce-0d6b19e76b97/jobs/16194).

## How urgent is a code review?

Medium priority. By the end-of-the week would be nice, as if this works I could then use a similar approach to avoid similar failures related to other validations.

jinjagit committed:
Add test: Update phase with ideas to a poll

Currently fails with error:
  ActiveRecord::RecordInvalid:
    Validation failed: Ideas phases is invalid
Passes if ideas_phase phase_is_ideation validation is disabled.

jinjagit committed:
Better test: models actual behaviour more closely

Now testing validity of a related IdeasPhase after updating an
an ideation phase with ideas to a poll phase.
Currently failing as expected, with an invalid IdeaPhase after
patch request, as 'fix' not yet implemented.

jinjagit commented:
Now testing validity of a related `IdeasPhase` after updating an ideation phase with ideas to a poll phase.

Currently failing as expected, with an invalid `IdeasPhase` after patch request, when we `expect(@project.ideas[0].ideas_phases.first.valid?).to eq true` as 'fix' not yet implemented.:
```
Failures:

  1) Phases when authenticated PATCH web_api/v1/phases/:id When updating ideation phase with ideas to a poll phase Existing related ideas_phase remains valid
     Failure/Error: expect(@project.ideas[0].ideas_phases.first.valid?).to eq true
     
       expected: true
            got: false
     
       (compared using ==)
     
       Diff:
       @@ -1,2 +1,2 @@
       -true
       +false
       
     # ./spec/acceptance/phases_spec.rb:294:in `block (5 levels) in <top (required)>'
     # /usr/local/bundle/gems/webmock-3.12.1/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'
```

jinjagit committed:
Comment out phase_is_ideation validation and check CI tests

This is way faster than running full test suite locally.

jinjagit committed:
Set expect block to test related ideas_phase still valid

This is the desired (green) outcome after commenting out the
related validation.

jinjagit committed:
Remove phase_is_ideation validation from IdeasPhase model

jinjagit committed:
Remove irrelevant let declaration from test

jinjagit commented:
After removing the `phase_is_ideation` validation, the new test now passes, asserting that a related `IdeasPhase` remains valid after the phase with ideas is updated to a non-ideation phase type, (after checking it fails when we assert the opposite: that it becomes invalid).

sebastienhoorens commented on back/spec/acceptance/phases_spec.rb:
```diff
@@ -274,6 +274,25 @@
+
+      describe "When updating ideation phase with ideas to a poll phase" do
+        before do
+          @project.phases.first.update(
```
It's better to use `update!` instead of `update`. `update` does not throw an error on failure, which gives two potential issues:
1. The test may not do what you want.
2. When trying to fix the test, it can take a long time to realize where things are going wrong.

sebastienhoorens commented on back/spec/acceptance/phases_spec.rb:
```diff
@@ -274,6 +274,25 @@
+        let(:participation_method) { 'poll' }
+
+        example "Existing related ideas_phase remains valid" do
+          expect(@project.ideas[0].ideas_phases.first.valid?).to eq true
```
`phase` is already there (which is what the spec is using to find out which phase to update). There should be no need to do e.g. `@project.ideas[0].ideas_phases.first`, you can just replace it by `phase`. But after `do_request` you might need to do `phase.reload` to tell Rails to get the latest values (the new participation method) from the DB.

jinjagit replied:
@sebastienhoorens I think `phase` and `ideas_phase` are not the same thing. We are checking the validity of `ideas_phase` which is a through-table entry associated with the idea (linking an `idea_id` to a specific `phase_id`).  

Replacing `@project.ideas[0].ideas_phases.first` with `phase` means the test will pass even when the validation is still applied.

jinjagit replied:
I _can_ replace occurrences of `@project.phases.first` with `phase`, however.

Maybe I can also find a way to get a related `idea_phase` through a `phase` too, without needing to reference `@project.ideas`.

jinjagit replied:
I have gone with using `let(:ideas_phase) { phase.ideas[0].ideas_phases.first }` to create a reference to a specific related `IdeasPhase` instance.

sebastienhoorens replied:
Oh right, sorry! Introducing an `ideas_phase` variable is indeed a good solution :)

sebastienhoorens reviewed (changes requested):
You'll also need to write a script (rake task) to fix the existing data.

jinjagit committed:
Improve test (see messge details)

Remove unnecessary let(:id) statement
Use phase in place of @project.phases.first
Define let(:ideas_phase) to to create a reference to specific
IdeasPhase instance.

jinjagit committed:
Use update! in test

jinjagit commented:
> You'll also need to write a script (rake task) to fix the existing data.

@sebastienhoorens I am not sure I understand...   

I assume the new rake task would live in `back/engines/ee/multi_tenancy/lib/tasks/core/fix_inconsistent_data.rake`  

Since we did not have to change any data to (hopefully) pass the `ideas_phase` inconsistent data check (we just removed the validation that caused the failures), I'm not sure what a rake task would be used for (perhaps to re-run the inconsistent data-check against existing `ideas_phase` records?)  

If we, indeed, do need a rake task, then some advice on the form would be gratefully received. (I couldn't see any obvious existing task to use as a template). Also some ideas on how to test the task locally would really help.

sebastienhoorens commented:
> > You'll also need to write a script (rake task) to fix the existing data.
> 
> @sebastienhoorens I am not sure I understand...
> 
> I assume the new rake task would live in `back/engines/ee/multi_tenancy/lib/tasks/core/fix_inconsistent_data.rake`
> 
> Since we did not have to change any data to (hopefully) pass the `ideas_phase` inconsistent data check (we just removed the validation that caused the failures), I'm not sure what a rake task would be used for (perhaps to re-run the inconsistent data-check against existing `ideas_phase` records?)
> 
> If we, indeed, do need a rake task, then some advice on the form would be gratefully received. (I couldn't see any obvious existing task to use as a template). Also some ideas on how to test the task locally would really help.

You're right! No migration of the existing data is needed in this case, good to go! :)
